IPs/RSKIP178.md

```diff
@@ -0,0 +1,375 @@
+	// * external uncle anchor heavyblock
+    // * mainchain anchor heavyblock
+    includeInBtcBlockReferences = true
+	if (fromHeavyblocksArray)
+		// external uncle anchor heavyblock
+		immediateDifficultyContribution = C(RSKblockHeader,bitcoinMergedMiningHeader)
+	else
+		// internal anchor heavyblock or uncle anchor heavyblock
+		immediateDifficultyContribution = A(rskBlockHeader,bitcoinMergedMiningHeader)
+else
+	// elements in fromHeavyblocksArray must be blocks solved at Bitcoin difficulty 
+	if fromHeavyblocksArray return
+	
+isValid = true
+return
+```
+
+In the shown pseudo-code *btcProofOfWorkValid()* checks that the proof of work of the Bitcoin header matches the difficulty specified in the Bitcoin header. The algorithm computes the *btcDifficulty0* specified by Bitcoin header and checks that it is higher than the difficulty of the RSK block that includes it. The call to the function *getMMInfo()* checks that the *bitcoinMergedMiningMerkleProof* is correct in relation with the other two fields if the mergeMiningProof, according to the same consensus rules used to validate merge-mining.
+
+Every merge-mining proof,  in mainchain, in the uncles list or in the heavyblocks list must be validated with the *getPoWInfo()* method. If the method returns isValid==false, then the block is invalid. If the method returns *iincludeInBtcBlockReferences*==true, then this means it is an anchor or confirmation heavyblock and therefore the associated Bitcoin block hash is must be included to the list *btcBlockReferences* that can be referenced by *btcBlockPreviouslyIncluded()*. 
+
+The function *btcBlockPreviouslyIncluded()* returns true if the bitcoin header with the hash given was included in the  *btcBlockReferences*  list. This array maintains a list of all included heavyblocks in the past 256 blocks. The array need not to be stored, as it can be recomputed each time from the heavyblocks, uncles and mainchain blocks in the past 256 blocks. Note that references may form a DAG, and therefore an included heavyblock may have more than one child in the last 256 blocks. 
+
+### Cumulative Difficulty
+
+Let's call *A* to the difficulty contributed by anchor heavyblocks to the chain cumulative difficulty. 
+
+Let's call *R* to the difficulty contributed by normal RSK blocks whose Bitcoin header difficulty has not reached the *btcDifficulty0*.
+
+Let's call *C* to the difficulty  contributed by confirmation heavyblocks to the chain cumulative difficulty.
+
+A, R, C are functions of the current RSK block header and the Bitcoin block header being referenced by the merge-mining fields.
+
+We propose the following functions:
+
+*A=btcDifficulty0/2*
+
+*R=difficulty/2*
+
+*C=btcDifficulty0*
+
+These formulas require using 256-bit arithmetic. This assignment overstates on average the RSK blockchain cumulative difficulty by at most 1/20 (the ratios of average block intervals between RSK and Bitcoin), which we consider too low to use the exact formulas.
```

Comment:
These formulas require the use of 256-bit arithmetic. This assignment overstates on average the RSK blockchain cumulative difficulty by at most 1/20 (the ratios of average block intervals between RSK and Bitcoin), which we consider too low to use the exact formulas.